pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -387,129 +525,82 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
-	}
-
-	return nodeChoice
-}
-
-func (c *Cluster) assignMons(mons []*monConfig) error {
-
-	// retrieve the set of cluster nodes and their monitor usage info
-	nodeZones, err := c.getNodeMonUsage()
-	if err != nil {
-		return fmt.Errorf("failed to get node monitor usage. %+v", err)
-	}
+	}()
 
-	// ensure all monitors have a node assignment. note that this isn't
-	// necessarily optimal: it does not try to move existing monitors which is
-	// handled by the periodic monitor health checks.
+	// ensure that all monitors have either (1) a node assignment that will be
+	// enforced using a node selector, or (2) configuration permits k8s to handle
+	// scheduling for the monitor.
 	for _, mon := range mons {
 
-		// monitor is already assigned to a node. nothing to do
+		// scheduling for this monitor has already been completed
 		if _, ok := c.mapping.Node[mon.DaemonName]; ok {
-			logger.Infof("assignmon: mon %s already assigned to a node", mon.DaemonName)
+			logger.Debugf("assignmon: mon %s already scheduled", mon.DaemonName)
 			continue
 		}
 
-		nodeChoice := scheduleMonitor(mon, nodeZones)
+		// determine a placement for the monitor. note that this scheduling is
+		// performed even when a node selector is not required. this may be
+		// non-optimal, but it is convenient to catch some failures early,
+		// before a decision is stored in the node mapping.
+		result, err := scheduleMonitor(c, mon)
 
-		if nodeChoice == nil {
-			return fmt.Errorf("assignmon: no valid nodes available for mon placement")
-		}
+		// even if an error occurs cleanup may be necessary
+		canaryCleanup = append(canaryCleanup, result)
 
-		// note that we do not need to worry about a false negative here (i.e. a
-		// node exists that _would_ pass this check) due to the search landing in a
-		// local minima because if a node had existed with a mon count of zero,
-		// scheduleMonitor would have chosen it over any node with a positive
-		// mon count.
-		if nodeChoice.MonCount > 0 && !c.spec.Mon.AllowMultiplePerNode {
-			return fmt.Errorf("assignmon: no empty nodes available for mon placement")
+		if err != nil {
+			return fmt.Errorf("assignmon: error scheduling monitor: %+v", err)
 		}
 
-		// make this decision visible when scheduling the next monitor
-		nodeChoice.MonCount++
-
-		logger.Infof("assignmon: mon %s assigned to node %s", mon.DaemonName, nodeChoice.Node.Name)
+		nodeChoice := result.Node
+		if nodeChoice == nil {
+			return fmt.Errorf("assignmon: could not schedule monitor %s", mon.DaemonName)
+		}
 
-		nodeInfo, err := getNodeInfoFromNode(*nodeChoice.Node)
-		if err != nil {
-			return fmt.Errorf("couldn't get node info from node %s. %+v",
-				nodeChoice.Node.Name, err)
+		// store nil in the node mapping to indicate that an explicit node
+		// placement is not being made. otherwise, the node choice will map
+		// directly to a node selector on the monitor pod.
+		var nodeInfo *NodeInfo = nil
+		if c.HostNetwork || c.spec.Mon.VolumeClaimTemplate == nil {
```

Comment:
If #3576 goes in first we'll need to rebase  and fix the change to the HostNetwork field.